split-llvm-extract.cpp

```diff
@@ -59,42 +62,82 @@ llvm::cl::opt<bool> dry("d",
 
 llvm::cl::opt<bool> verbose("v", llvm::cl::desc("Enable vrbose output."));
 
-struct MyPass : public llvm::InstVisitor<MyPass> {
+std::unordered_set<llvm::GlobalVariable *> iterateOperands(llvm::User& user) {
+  std::stack<llvm::Value *> operands;
   std::unordered_set<llvm::GlobalVariable *> globals;
 
-  void visitInstruction(llvm::Instruction &instruction) {
-    std::stack<llvm::Value *> operands;
+  for (uint32_t i = 0; i < user.getNumOperands(); i++) {
+    operands.push(user.getOperand(i));
+  }
+
+  while (operands.size() > 0) {
+    auto currentOperand = operands.top();
+    operands.pop();
+
+    if (currentOperand == nullptr) {
+      // TODO: ignore null's for now.
```

Comment:
Should we `err` here?

Reply:
The truth is I don't know, in practice it is fine some operand to be null, as llvm's instructions can have optional operands. So I am not sure what exactly to do.

Reply:
If in doubt always `err`, that way as soon as we hit a problem we're notified. With a concrete example it then becomes much clearer as to what we should do. As this suggests, I'm allergic to "do nothing because we don't know" code :)

Reply:
The problem is that everything breaks when I don't ignore null, because there are everywhere,

Reply:
I fixed it in: 0534b18 it turns out that what I was doing wasn't quite the best way to iterate the operands. The proper way wasn't working for me before but it works now. I think I had misused it before and decided to do it like that. But now it is working fine.